console/v1/0000_10_consolesample.crd.yaml

```diff
@@ -0,0 +1,121 @@
+                  type: string
+                  maxLength: 20
+                description:
+                  description: "description of the sample. A README.md-like content for additional information, links, pre and post conditions and instrucations. \n It is required and is shown on the details page. It is rendered as Markdown so that it can contain line breaks, links, and other simple formatting."
+                  type: string
+                icon:
+                  description: icon is an optional base64 encoded image that will be displayed beside the sample title. The icon should be an vector image for easy scaling. It will be shown on a white background (also when the dark theme is used) and in different sizes (currently not bigger then 40x100px). Most icons will be quadratic but the web console ensures that all aspect ratios work.
+                  type: string
+                provider:
+                  description: "provider like \"Red Hat\" is optional and shown with the prefix \"Provided by \". \n Currently, is only shown on the sample card tile below the title. \n It is optional and must be no more than 20 characters in length."
+                  type: string
+                  maxLength: 20
+                source:
+                  description: The actual sample source, like an external git repository or container image.
+                  type: object
+                  required:
+                    - type
+                  properties:
+                    containerImport:
+                      description: containerimport let the user import a container image.
+                      type: object
+                      required:
+                        - image
+                      properties:
+                        image:
+                          description: Reference to a container image like openshift/hello-openshift, docker.io/openshift/hello-openshift or quay.io/<username>/<image-name>
+                          type: string
+                          minLength: 1
+                    gitImport:
+                      description: gitimport let the user import code from a git repository.
+                      type: object
+                      required:
+                        - url
+                      properties:
+                        ref:
+                          description: Optional branch, tag, or commit.
+                          type: string
+                        url:
+                          description: Git repository URL.
+                          type: string
+                          minLength: 1
+                    type:
+                      description: 'type of the sample, currently supported: "GitImport";"ContainerImport"'
```

Comment:
Would it be possible to have a third option, something like `ProcessImport` so that templates could be generated dynamically? The `func` CLI - which is included as a binary in the Serverless distribution generates samples directly that are built into the binary. For example, `kn func create -l go -t cloudevents -p /source/fn` will create a new Golang function project that handles cloudevents in the `/source/fn` directory.

There are at least a couple of benefits to this approach. First, we are guaranteed to create a function that we know is compatible with the installed Serverless version. Also, this ensures that templates are kept up to date with library upgrades, etc. as this is already managed in the knative/func repository. Having duplicate templates in a separate examples repository would create additional maintenance work that duplicates that in the func repo.

Reply:
I don't see a straightforward option to run a cli command from the web console. Esp. when the cli creates some files. In theory, we could start a Pod with a volume. But I don't get how this will help the user to understand how the console can import, build and deploy a kn func.

For more complex scenarios, it might be an option to use OpenShift Templates.

But the idea of our (console) samples is to explain to the users how the console works and, in our case, how the "Import a Serverless function" feature works.

Reply:
We (the web console team) don't run `git clone` our own.

We "do this" by creating OpenShift `BuildConfigs` or Tekton `Pipelines` (and in the future hopefully Shipwright `Builds` for the user. They of course clone the repo, build a container image and push it into our internal image registry.

The web console also creates Deployments or a Knative `Service` to start these images for you.

Of course, we could create resources, that run `kn func && kn deploy` (as Pod, or Job) to archive something similar. But that's really different from the samples we have already.